<!doctype html><html lang=en><head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=robots content="index, follow"><link rel="icon shortcut" href=/obitools4-doc/favicon.ico sizes=32x32><link rel=icon href=/obitools4-doc/favicon.svg type=image/svg+xml><link rel=icon href=/obitools4-doc/favicon-dark.svg type=image/svg+xml media="(prefers-color-scheme: dark)"><link rel=icon href=/obitools4-doc/favicon-16x16.png type=image/png sizes=16x16><link rel=icon href=/obitools4-doc/favicon-32x32.png type=image/png sizes=32x32><link rel=apple-touch-icon href=/obitools4-doc/apple-touch-icon.png sizes=180x180><link fetchpriority=low href=/obitools4-doc/site.webmanifest rel=manifest><title>Obipcr: the electronic PCR tool – OBITools4 documentation</title>
<meta name=description content="Description : obipcr is the successor of ecoPCR. It is known as an electronic PCR software.
Synopsis obipcr --forward <string> --max-length|-L <int> --reverse <string> [--allowed-mismatches|-e <int>] [--batch-size <int>] [--circular|-c] [--compress|-Z] [--debug] [--delta|-D <int>] [--ecopcr] [--embl] [--fasta] [--fasta-output] [--fastq] [--fastq-output] [--force-one-cpu] [--fragmented] [--genbank] [--help|-h|-?] [--input-OBI-header] [--input-json-header] [--json-output] [--max-cpu <int>] [--min-length|-l <int>] [--no-order] [--no-progressbar] [--only-complete-flanking] [--out|-o <FILENAME>] [--output-OBI-header|-O] [--output-json-header] [--paired-with <FILENAME>] [--pprof] [--pprof-goroutine <int>] [--pprof-mutex <int>] [--skip-empty] [--solexa] [--version] [<args>] Options obipcr mandatory options: --forward: The forward primer used for the electronic PCR. IUPAC codes can be used in the pattern. --reverse: The reverse primer used for the electronic PCR. IUPAC codes can be used in the pattern. --max-length | -L : Maximum length of the barcode, primers excluded. Other obipcr specific options: --allowed-mismatches | -e : Maximum number of mismatches allowed for each primer (default: 0). --min-length | -l : Minimum length of the barcode primers excluded (default: no minimum length). --circular | -c : Considers that sequences are circular. (default: sequences are considered linear) --delta | -D : Without this option, only the barcode sequences will be output, without the priming sites. This option allows to add the priming sites and the flanking sequences of the priming sites over a length of delta to each side of the barcode. --only-complete-flanking: Works in conjunction with –delta. Prints only sequences with full-length flanking sequences (default: prints every sequence regardless of whether the flanking sequences are present). Controlling the input data --fasta: Indicate that sequence data is following the FASTA format. --fastq: Indicate that sequence data is following the FASTQ format. --embl: Indicate that sequence data is following the EMBL flatfile format. --genbank: Indicate that sequence data is following the Genbank flatfile format. --ecopcr: Indicate that sequence data is following old ecoPCR tabulated format. --input-OBI-header: FASTA/FASTQ title line annotations follow the old OBI format. --input-json-header: FASTA/FASTQ title line annotations follow the json format. --solexa: Decodes quality string according to the Solexa specification. (default: the standard Sanger encoding is used, env: OBISSOLEXA) Controlling the output data --compress | -Z : Output is compressed using gzip (default: false). --no-order: The OBITools ensure that the order between the input file and the output file does not change. When multiple files are processed, they are processed one at a time. If the –no-order option is added to a command, multiple input files can be opened at the same time and their contents processed in parallel. This usually increases processing speed, but does not guarantee the order of the sequences in the output file. Also, processing multiple files in parallel may require more memory to perform the computation. --fasta-output: Write sequence in fasta format (default if no quality data available). --fastq-output: Write sequence in fastq format (default if quality data available). --json-output: Write sequence in json format. --out | -o <FILENAME>: Filename used for saving the output (default: “-”, the standard output) --output-OBI-header | -O : output FASTA/FASTQ title line annotations follow OBI format (default: JSON). --output-json-header: output FASTA/FASTQ title line annotations follow json format (the default format). --skip-empty: Sequences of length equal to zero are suppressed from the output (default: false). --no-progressbar: Disable the progress bar printing (default: false). General options --help | -h|-? : Show this help. --version: Print the version and exits. Computation related options --max-cpu <INTEGER>: OBITools can take advantage of your computer’s multi-core architecture by parallelizing the computation across all available CPUs. Computing on more CPUs usually requires more memory to perform the computation. Reducing the number of CPUs used to perform a calculation is also a way to indirectly control the amount of memory used by the process. The number of CPUs used by OBITools can also be controlled by setting the OBIMAXCPU environment variable. --force-one-cpu: Force to use only one CPU core for parallel processing (default: false). --batch-size <INTEGER>: Number of sequence per batch for parallel processing (default: 1000, env: OBIBATCHSIZE) Debug related options --debug: Enable debug mode, by setting log level to debug. (default: false, env: OBIDEBUG). --pprof: Enable pprof server. Look at the log for details. (default: false). --pprof-mutex <INTEGER>: Enable profiling of mutex lock. (default: 10, env: OBIPPROFMUTEX) --pprof-goroutine <INTEGER>: Enable profiling of goroutine blocking profile. (default: 6060, env: OBIPPROFGOROUTINE) Examples The minimal ecopcr command looks like this:"><link rel=canonical href=https://metabarcoding.github.io/obitools4-doc/commands/alignments/obipcr/ itemprop=url><meta property="og:title" content="Obipcr: the electronic PCR tool"><meta property="og:description" content="Description :
    obipcr is the successor of ecoPCR. It is known as an electronic PCR software.
Synopsis
    
  

obipcr --forward <string> --max-length|-L <int> --reverse <string>
       [--allowed-mismatches|-e <int>] [--batch-size <int>] [--circular|-c]
       [--compress|-Z] [--debug] [--delta|-D <int>] [--ecopcr] [--embl]
       [--fasta] [--fasta-output] [--fastq] [--fastq-output]
       [--force-one-cpu] [--fragmented] [--genbank] [--help|-h|-?]
       [--input-OBI-header] [--input-json-header] [--json-output]
       [--max-cpu <int>] [--min-length|-l <int>] [--no-order]
       [--no-progressbar] [--only-complete-flanking] [--out|-o <FILENAME>]
       [--output-OBI-header|-O] [--output-json-header]
       [--paired-with <FILENAME>] [--pprof] [--pprof-goroutine <int>]
       [--pprof-mutex <int>] [--skip-empty] [--solexa] [--version] [<args>]
  
    
    
  


  

Options
    obipcr mandatory options:
    

--forward: The forward primer used for the electronic PCR. IUPAC codes can be used in the pattern.



--reverse: The reverse primer used for the electronic PCR. IUPAC codes can be used in the pattern.



--max-length
| -L
: Maximum length of the barcode, primers excluded.



Other obipcr specific options:
    

--allowed-mismatches
| -e
: Maximum number of mismatches allowed for each primer (default: 0).



--min-length
| -l
: Minimum length of the barcode primers excluded (default: no minimum length).



--circular
| -c
: Considers that sequences are circular. (default: sequences are considered linear)



--delta
| -D
: Without this option, only the barcode sequences will be output, without the priming sites. This option allows to add the priming sites and the flanking sequences of the priming sites over a length of delta to each side of the barcode.



--only-complete-flanking: Works in conjunction with &ndash;delta. Prints only sequences with full-length flanking sequences (default: prints every sequence regardless of whether the flanking sequences are present).



Controlling the input data
    


    --fasta: Indicate that sequence data is following the FASTA format.
    --fastq: Indicate that sequence data is following the FASTQ format.
    --embl: Indicate that sequence data is following the EMBL flatfile format.
    --genbank: Indicate that sequence data is following the Genbank flatfile format.
    --ecopcr: Indicate that sequence data is following old ecoPCR tabulated format.
    --input-OBI-header: FASTA/FASTQ title line annotations follow the old OBI format.
    --input-json-header: FASTA/FASTQ title line annotations follow the json format.
    --solexa: Decodes quality string according to the Solexa specification. (default: the standard Sanger encoding is used, env: OBISSOLEXA)


Controlling the output data
    


    --compress
| -Z
: Output is compressed using gzip (default: false).
    --no-order: The OBITools ensure that the order between the input file and
the output file does not change. When multiple files are processed,
they are processed one at a time.
If the &ndash;no-order option is added to a command, multiple input
files can be opened at the same time and their contents processed
in parallel. This usually increases processing speed, but does not
guarantee the order of the sequences in the output file.
Also, processing multiple files in parallel may require more memory
to perform the computation.
    --fasta-output: Write sequence in fasta format (default if no quality data available).
    --fastq-output: Write sequence in fastq format (default if quality data available).
    --json-output: Write sequence in json format.
    --out
| -o
 <FILENAME>: Filename used for saving the output (default: &ldquo;-&rdquo;, the standard output)
    --output-OBI-header
| -O
: output FASTA/FASTQ title line annotations follow OBI format (default: JSON).
    --output-json-header: output FASTA/FASTQ title line annotations follow json format (the default format).
    --skip-empty: Sequences of length equal to zero are suppressed from the output (default: false).
    --no-progressbar: Disable the progress bar printing (default: false).

General options
    


    --help
| -h|-?
: Show this help.
    --version: Print the version and exits.




Computation related options
    


    --max-cpu <INTEGER>: OBITools can take advantage of your computer&rsquo;s multi-core
architecture by parallelizing the computation across all available CPUs.
Computing on more CPUs usually requires more memory to perform the
computation. Reducing the number of CPUs used to perform a calculation
is also a way to indirectly control the amount of memory used by the
process. The number of CPUs used by OBITools can also be controlled
by setting the OBIMAXCPU environment variable.
    --force-one-cpu: Force to use only one CPU core for parallel processing (default: false).
    --batch-size <INTEGER>: Number of sequence per batch for parallel processing (default: 1000, env: OBIBATCHSIZE)


Debug related options
    


    --debug: Enable debug mode, by setting log level to debug. (default: false, env: OBIDEBUG).
    --pprof: Enable pprof server. Look at the log for details. (default: false).
    --pprof-mutex <INTEGER>: Enable profiling of mutex lock. (default: 10, env: OBIPPROFMUTEX)
    --pprof-goroutine <INTEGER>: Enable profiling of goroutine blocking profile. (default: 6060, env: OBIPPROFGOROUTINE)

Examples
    The minimal ecopcr command looks like this:"><meta property="og:type" content="article"><meta property="og:url" content="https://metabarcoding.github.io/obitools4-doc/commands/alignments/obipcr/"><meta property="article:section" content="commands"><meta itemprop=name content="Obipcr: the electronic PCR tool"><meta itemprop=description content="Description : obipcr is the successor of ecoPCR. It is known as an electronic PCR software.
Synopsis obipcr --forward <string> --max-length|-L <int> --reverse <string> [--allowed-mismatches|-e <int>] [--batch-size <int>] [--circular|-c] [--compress|-Z] [--debug] [--delta|-D <int>] [--ecopcr] [--embl] [--fasta] [--fasta-output] [--fastq] [--fastq-output] [--force-one-cpu] [--fragmented] [--genbank] [--help|-h|-?] [--input-OBI-header] [--input-json-header] [--json-output] [--max-cpu <int>] [--min-length|-l <int>] [--no-order] [--no-progressbar] [--only-complete-flanking] [--out|-o <FILENAME>] [--output-OBI-header|-O] [--output-json-header] [--paired-with <FILENAME>] [--pprof] [--pprof-goroutine <int>] [--pprof-mutex <int>] [--skip-empty] [--solexa] [--version] [<args>] Options obipcr mandatory options: --forward: The forward primer used for the electronic PCR. IUPAC codes can be used in the pattern. --reverse: The reverse primer used for the electronic PCR. IUPAC codes can be used in the pattern. --max-length | -L : Maximum length of the barcode, primers excluded. Other obipcr specific options: --allowed-mismatches | -e : Maximum number of mismatches allowed for each primer (default: 0). --min-length | -l : Minimum length of the barcode primers excluded (default: no minimum length). --circular | -c : Considers that sequences are circular. (default: sequences are considered linear) --delta | -D : Without this option, only the barcode sequences will be output, without the priming sites. This option allows to add the priming sites and the flanking sequences of the priming sites over a length of delta to each side of the barcode. --only-complete-flanking: Works in conjunction with –delta. Prints only sequences with full-length flanking sequences (default: prints every sequence regardless of whether the flanking sequences are present). Controlling the input data --fasta: Indicate that sequence data is following the FASTA format. --fastq: Indicate that sequence data is following the FASTQ format. --embl: Indicate that sequence data is following the EMBL flatfile format. --genbank: Indicate that sequence data is following the Genbank flatfile format. --ecopcr: Indicate that sequence data is following old ecoPCR tabulated format. --input-OBI-header: FASTA/FASTQ title line annotations follow the old OBI format. --input-json-header: FASTA/FASTQ title line annotations follow the json format. --solexa: Decodes quality string according to the Solexa specification. (default: the standard Sanger encoding is used, env: OBISSOLEXA) Controlling the output data --compress | -Z : Output is compressed using gzip (default: false). --no-order: The OBITools ensure that the order between the input file and the output file does not change. When multiple files are processed, they are processed one at a time. If the –no-order option is added to a command, multiple input files can be opened at the same time and their contents processed in parallel. This usually increases processing speed, but does not guarantee the order of the sequences in the output file. Also, processing multiple files in parallel may require more memory to perform the computation. --fasta-output: Write sequence in fasta format (default if no quality data available). --fastq-output: Write sequence in fastq format (default if quality data available). --json-output: Write sequence in json format. --out | -o <FILENAME>: Filename used for saving the output (default: “-”, the standard output) --output-OBI-header | -O : output FASTA/FASTQ title line annotations follow OBI format (default: JSON). --output-json-header: output FASTA/FASTQ title line annotations follow json format (the default format). --skip-empty: Sequences of length equal to zero are suppressed from the output (default: false). --no-progressbar: Disable the progress bar printing (default: false). General options --help | -h|-? : Show this help. --version: Print the version and exits. Computation related options --max-cpu <INTEGER>: OBITools can take advantage of your computer’s multi-core architecture by parallelizing the computation across all available CPUs. Computing on more CPUs usually requires more memory to perform the computation. Reducing the number of CPUs used to perform a calculation is also a way to indirectly control the amount of memory used by the process. The number of CPUs used by OBITools can also be controlled by setting the OBIMAXCPU environment variable. --force-one-cpu: Force to use only one CPU core for parallel processing (default: false). --batch-size <INTEGER>: Number of sequence per batch for parallel processing (default: 1000, env: OBIBATCHSIZE) Debug related options --debug: Enable debug mode, by setting log level to debug. (default: false, env: OBIDEBUG). --pprof: Enable pprof server. Look at the log for details. (default: false). --pprof-mutex <INTEGER>: Enable profiling of mutex lock. (default: 10, env: OBIPPROFMUTEX) --pprof-goroutine <INTEGER>: Enable profiling of goroutine blocking profile. (default: 6060, env: OBIPPROFGOROUTINE) Examples The minimal ecopcr command looks like this:"><meta itemprop=wordCount content="852"><meta name=twitter:card content="summary"><meta name=twitter:title content="Obipcr: the electronic PCR tool"><meta name=twitter:description content="Description : obipcr is the successor of ecoPCR. It is known as an electronic PCR software.
Synopsis obipcr --forward <string> --max-length|-L <int> --reverse <string> [--allowed-mismatches|-e <int>] [--batch-size <int>] [--circular|-c] [--compress|-Z] [--debug] [--delta|-D <int>] [--ecopcr] [--embl] [--fasta] [--fasta-output] [--fastq] [--fastq-output] [--force-one-cpu] [--fragmented] [--genbank] [--help|-h|-?] [--input-OBI-header] [--input-json-header] [--json-output] [--max-cpu <int>] [--min-length|-l <int>] [--no-order] [--no-progressbar] [--only-complete-flanking] [--out|-o <FILENAME>] [--output-OBI-header|-O] [--output-json-header] [--paired-with <FILENAME>] [--pprof] [--pprof-goroutine <int>] [--pprof-mutex <int>] [--skip-empty] [--solexa] [--version] [<args>] Options obipcr mandatory options: --forward: The forward primer used for the electronic PCR. IUPAC codes can be used in the pattern. --reverse: The reverse primer used for the electronic PCR. IUPAC codes can be used in the pattern. --max-length | -L : Maximum length of the barcode, primers excluded. Other obipcr specific options: --allowed-mismatches | -e : Maximum number of mismatches allowed for each primer (default: 0). --min-length | -l : Minimum length of the barcode primers excluded (default: no minimum length). --circular | -c : Considers that sequences are circular. (default: sequences are considered linear) --delta | -D : Without this option, only the barcode sequences will be output, without the priming sites. This option allows to add the priming sites and the flanking sequences of the priming sites over a length of delta to each side of the barcode. --only-complete-flanking: Works in conjunction with –delta. Prints only sequences with full-length flanking sequences (default: prints every sequence regardless of whether the flanking sequences are present). Controlling the input data --fasta: Indicate that sequence data is following the FASTA format. --fastq: Indicate that sequence data is following the FASTQ format. --embl: Indicate that sequence data is following the EMBL flatfile format. --genbank: Indicate that sequence data is following the Genbank flatfile format. --ecopcr: Indicate that sequence data is following old ecoPCR tabulated format. --input-OBI-header: FASTA/FASTQ title line annotations follow the old OBI format. --input-json-header: FASTA/FASTQ title line annotations follow the json format. --solexa: Decodes quality string according to the Solexa specification. (default: the standard Sanger encoding is used, env: OBISSOLEXA) Controlling the output data --compress | -Z : Output is compressed using gzip (default: false). --no-order: The OBITools ensure that the order between the input file and the output file does not change. When multiple files are processed, they are processed one at a time. If the –no-order option is added to a command, multiple input files can be opened at the same time and their contents processed in parallel. This usually increases processing speed, but does not guarantee the order of the sequences in the output file. Also, processing multiple files in parallel may require more memory to perform the computation. --fasta-output: Write sequence in fasta format (default if no quality data available). --fastq-output: Write sequence in fastq format (default if quality data available). --json-output: Write sequence in json format. --out | -o <FILENAME>: Filename used for saving the output (default: “-”, the standard output) --output-OBI-header | -O : output FASTA/FASTQ title line annotations follow OBI format (default: JSON). --output-json-header: output FASTA/FASTQ title line annotations follow json format (the default format). --skip-empty: Sequences of length equal to zero are suppressed from the output (default: false). --no-progressbar: Disable the progress bar printing (default: false). General options --help | -h|-? : Show this help. --version: Print the version and exits. Computation related options --max-cpu <INTEGER>: OBITools can take advantage of your computer’s multi-core architecture by parallelizing the computation across all available CPUs. Computing on more CPUs usually requires more memory to perform the computation. Reducing the number of CPUs used to perform a calculation is also a way to indirectly control the amount of memory used by the process. The number of CPUs used by OBITools can also be controlled by setting the OBIMAXCPU environment variable. --force-one-cpu: Force to use only one CPU core for parallel processing (default: false). --batch-size <INTEGER>: Number of sequence per batch for parallel processing (default: 1000, env: OBIBATCHSIZE) Debug related options --debug: Enable debug mode, by setting log level to debug. (default: false, env: OBIDEBUG). --pprof: Enable pprof server. Look at the log for details. (default: false). --pprof-mutex <INTEGER>: Enable profiling of mutex lock. (default: 10, env: OBIPPROFMUTEX) --pprof-goroutine <INTEGER>: Enable profiling of goroutine blocking profile. (default: 6060, env: OBIPPROFGOROUTINE) Examples The minimal ecopcr command looks like this:"><link rel=preload href=/obitools4-doc/css/compiled/main.min.53c9617cf9cb82685ec5af463dec3ecab7f3b4c3ad5c9cf22185fc7cf9b39519.css as=style integrity="sha256-U8lhfPnLgmhexa9GPew+yrfztMOtXJzyIYX8fPmzlRk="><link href=/obitools4-doc/css/compiled/main.min.53c9617cf9cb82685ec5af463dec3ecab7f3b4c3ad5c9cf22185fc7cf9b39519.css rel=stylesheet integrity="sha256-U8lhfPnLgmhexa9GPew+yrfztMOtXJzyIYX8fPmzlRk="><link href=/obitools4-doc/css/custom.min.e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.css rel=stylesheet integrity="sha256-47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU="><script>const defaultTheme="system",setDarkTheme=()=>{document.documentElement.classList.add("dark"),document.documentElement.style.colorScheme="dark"},setLightTheme=()=>{document.documentElement.classList.remove("dark"),document.documentElement.style.colorScheme="light"};"color-theme"in localStorage?localStorage.getItem("color-theme")==="dark"?setDarkTheme():setLightTheme():(defaultTheme==="dark"?setDarkTheme():setLightTheme(),defaultTheme==="system"&&(window.matchMedia("(prefers-color-scheme: dark)").matches?setDarkTheme():setLightTheme()))</script></head><body dir=ltr><div class="nav-container hx-sticky hx-top-0 hx-z-20 hx-w-full hx-bg-transparent print:hx-hidden"><div class="nav-container-blur hx-pointer-events-none hx-absolute hx-z-[-1] hx-h-full hx-w-full hx-bg-white dark:hx-bg-dark hx-shadow-[0_2px_4px_rgba(0,0,0,.02),0_1px_0_rgba(0,0,0,.06)] contrast-more:hx-shadow-[0_0_0_1px_#000] dark:hx-shadow-[0_-1px_0_rgba(255,255,255,.1)_inset] contrast-more:dark:hx-shadow-[0_0_0_1px_#fff]"></div><nav class="hx-mx-auto hx-flex hx-items-center hx-justify-end hx-gap-2 hx-h-16 hx-px-6 hx-max-w-[90rem]"><a class="hx-flex hx-items-center hover:hx-opacity-75 ltr:hx-mr-auto rtl:hx-ml-auto" href=/obitools4-doc/><img class="hx-block dark:hx-hidden" src=/obitools4-doc/images/logo.svg alt="OBITools4 documentation" height=20 width=20>
<img class="hx-hidden dark:hx-block" src=/obitools4-doc/images/logo.svg alt="OBITools4 documentation" height=20 width=20>
<span class="hx-mx-2 hx-font-extrabold hx-inline hx-select-none" title="OBITools4 documentation">OBITools4 documentation</span>
</a><button type=button aria-label=Menu class="hamburger-menu -hx-mr-2 hx-rounded hx-p-2 active:hx-bg-gray-400/20 md:hx-hidden"><svg height="24" fill="none" viewBox="0 0 24 24" stroke="currentcolor"><g><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 8H20"/></g><g><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 16H20"/></g></svg></button></nav></div><div class='hx-mx-auto hx-flex hx-max-w-screen-xl'><div class="mobile-menu-overlay [transition:background-color_1.5s_ease] hx-fixed hx-inset-0 hx-z-10 hx-bg-black/80 dark:hx-bg-black/60 hx-hidden"></div><aside class="sidebar-container hx-flex hx-flex-col print:hx-hidden md:hx-top-16 md:hx-shrink-0 md:hx-w-64 md:hx-self-start max-md:[transform:translate3d(0,-100%,0)] md:hx-sticky"><div class="hx-px-4 hx-pt-4 md:hx-hidden"><div class="search-wrapper hx-relative md:hx-w-64"><div class="hx-relative hx-flex hx-items-center hx-text-gray-900 contrast-more:hx-text-gray-800 dark:hx-text-gray-300 contrast-more:dark:hx-text-gray-300"><input placeholder=Search... class="search-input hx-block hx-w-full hx-appearance-none hx-rounded-lg hx-px-3 hx-py-2 hx-transition-colors hx-text-base hx-leading-tight md:hx-text-sm hx-bg-black/[.05] dark:hx-bg-gray-50/10 focus:hx-bg-white dark:focus:hx-bg-dark placeholder:hx-text-gray-500 dark:placeholder:hx-text-gray-400 contrast-more:hx-border contrast-more:hx-border-current" type=search spellcheck=false>
<kbd class="hx-absolute hx-my-1.5 hx-select-none ltr:hx-right-1.5 rtl:hx-left-1.5 hx-h-5 hx-rounded hx-bg-white hx-px-1.5 hx-font-mono hx-text-[10px] hx-font-medium hx-text-gray-500 hx-border dark:hx-border-gray-100/20 dark:hx-bg-dark/50 contrast-more:hx-border-current contrast-more:hx-text-current contrast-more:dark:hx-border-current hx-items-center hx-gap-1 hx-transition-opacity hx-pointer-events-none hx-hidden sm:hx-flex">CTRL K</kbd></div><div><ul class="search-results hextra-scrollbar hx-hidden hx-border hx-border-gray-200 hx-bg-white hx-text-gray-100 dark:hx-border-neutral-800 dark:hx-bg-neutral-900 hx-absolute hx-top-full hx-z-20 hx-mt-2 hx-overflow-auto hx-overscroll-contain hx-rounded-xl hx-py-2.5 hx-shadow-xl hx-max-h-[min(calc(50vh-11rem-env(safe-area-inset-bottom)),400px)] md:hx-max-h-[min(calc(100vh-5rem-env(safe-area-inset-bottom)),400px)] hx-inset-x-0 ltr:md:hx-left-auto rtl:md:hx-right-auto contrast-more:hx-border contrast-more:hx-border-gray-900 contrast-more:dark:hx-border-gray-50 hx-w-screen hx-min-h-[100px] hx-max-w-[min(calc(100vw-2rem),calc(100%+20rem))]" style="transition:max-height .2s ease 0s"></ul></div></div></div><div class="hextra-scrollbar hx-overflow-y-auto hx-overflow-x-hidden hx-p-4 hx-grow md:hx-h-[calc(100vh-var(--navbar-height)-var(--menu-height))]"><ul class="hx-flex hx-flex-col hx-gap-1 md:hx-hidden"><li class=open><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/commands/>Commands
<span class=hextra-sidebar-collapsible-button><svg fill="none" viewBox="0 0 24 24" stroke="currentcolor" class="hx-h-[18px] hx-min-w-[18px] hx-rounded-sm hx-p-0.5 hover:hx-bg-gray-800/5 dark:hover:hx-bg-gray-100/5"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M9 5l7 7-7 7" class="hx-origin-center hx-transition-transform rtl:-hx-rotate-180"/></svg></span></a><div class="ltr:hx-pr-0 hx-overflow-hidden"><ul class='hx-relative hx-flex hx-flex-col hx-gap-1 before:hx-absolute before:hx-inset-y-1 before:hx-w-px before:hx-bg-gray-200 before:hx-content-[""] ltr:hx-ml-3 ltr:hx-pl-3 ltr:before:hx-left-0 rtl:hx-mr-3 rtl:hx-pr-3 rtl:before:hx-right-0 dark:before:hx-bg-neutral-800'><li class="hx-flex hx-flex-col"><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/commands/basics/>Basics
<span class=hextra-sidebar-collapsible-button><svg fill="none" viewBox="0 0 24 24" stroke="currentcolor" class="hx-h-[18px] hx-min-w-[18px] hx-rounded-sm hx-p-0.5 hover:hx-bg-gray-800/5 dark:hover:hx-bg-gray-100/5"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M9 5l7 7-7 7" class="hx-origin-center hx-transition-transform rtl:-hx-rotate-180"/></svg></span></a><div class="ltr:hx-pr-0 hx-overflow-hidden"><ul class='hx-relative hx-flex hx-flex-col hx-gap-1 before:hx-absolute before:hx-inset-y-1 before:hx-w-px before:hx-bg-gray-200 before:hx-content-[""] ltr:hx-ml-3 ltr:hx-pl-3 ltr:before:hx-left-0 rtl:hx-mr-3 rtl:hx-pr-3 rtl:before:hx-right-0 dark:before:hx-bg-neutral-800'><li class="hx-flex hx-flex-col"><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/commands/basics/obicount/>obicount: Counting sequence records</a></li></ul></div></li><li class="hx-flex hx-flex-col open"><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/commands/alignments/>Sequence alignments
<span class=hextra-sidebar-collapsible-button><svg fill="none" viewBox="0 0 24 24" stroke="currentcolor" class="hx-h-[18px] hx-min-w-[18px] hx-rounded-sm hx-p-0.5 hover:hx-bg-gray-800/5 dark:hover:hx-bg-gray-100/5"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M9 5l7 7-7 7" class="hx-origin-center hx-transition-transform rtl:-hx-rotate-180"/></svg></span></a><div class="ltr:hx-pr-0 hx-overflow-hidden"><ul class='hx-relative hx-flex hx-flex-col hx-gap-1 before:hx-absolute before:hx-inset-y-1 before:hx-w-px before:hx-bg-gray-200 before:hx-content-[""] ltr:hx-ml-3 ltr:hx-pl-3 ltr:before:hx-left-0 rtl:hx-mr-3 rtl:hx-pr-3 rtl:before:hx-right-0 dark:before:hx-bg-neutral-800'><li class="hx-flex hx-flex-col"><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/commands/alignments/obipairing/>Obipairing: align the forward and reverse paired reads</a></li><li class="hx-flex hx-flex-col open"><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
sidebar-active-item hx-bg-primary-100 hx-font-semibold hx-text-primary-800 contrast-more:hx-border contrast-more:hx-border-primary-500 dark:hx-bg-primary-400/10 dark:hx-text-primary-600 contrast-more:dark:hx-border-primary-500" href=/obitools4-doc/commands/alignments/obipcr/>Obipcr: the electronic PCR tool</a><ul class='hx-flex hx-flex-col hx-gap-1 hx-relative before:hx-absolute before:hx-inset-y-1 before:hx-w-px before:hx-bg-gray-200 before:hx-content-[""] dark:before:hx-bg-neutral-800 ltr:hx-pl-3 ltr:before:hx-left-0 rtl:hx-pr-3 rtl:before:hx-right-0 ltr:hx-ml-3 rtl:hx-mr-3'><li><a href=#description- class="hx-flex hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [word-break:break-word] hx-cursor-pointer [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] contrast-more:hx-border hx-gap-2 before:hx-opacity-25 before:hx-content-['#'] hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:hx-text-gray-900 contrast-more:dark:hx-text-gray-50 contrast-more:hx-border-transparent contrast-more:hover:hx-border-gray-900 contrast-more:dark:hover:hx-border-gray-50">Description :</a></li></ul></li></ul></div></li></ul></div></li><li><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/docs/>Docs</a></li></ul><ul class="hx-flex hx-flex-col hx-gap-1 max-md:hx-hidden"><li><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/commands/basics/>Basics
<span class=hextra-sidebar-collapsible-button><svg fill="none" viewBox="0 0 24 24" stroke="currentcolor" class="hx-h-[18px] hx-min-w-[18px] hx-rounded-sm hx-p-0.5 hover:hx-bg-gray-800/5 dark:hover:hx-bg-gray-100/5"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M9 5l7 7-7 7" class="hx-origin-center hx-transition-transform rtl:-hx-rotate-180"/></svg></span></a><div class="ltr:hx-pr-0 hx-overflow-hidden"><ul class='hx-relative hx-flex hx-flex-col hx-gap-1 before:hx-absolute before:hx-inset-y-1 before:hx-w-px before:hx-bg-gray-200 before:hx-content-[""] ltr:hx-ml-3 ltr:hx-pl-3 ltr:before:hx-left-0 rtl:hx-mr-3 rtl:hx-pr-3 rtl:before:hx-right-0 dark:before:hx-bg-neutral-800'><li class="hx-flex hx-flex-col"><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/commands/basics/obicount/>obicount: Counting sequence records</a></li></ul></div></li><li class=open><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/commands/alignments/>Sequence alignments
<span class=hextra-sidebar-collapsible-button><svg fill="none" viewBox="0 0 24 24" stroke="currentcolor" class="hx-h-[18px] hx-min-w-[18px] hx-rounded-sm hx-p-0.5 hover:hx-bg-gray-800/5 dark:hover:hx-bg-gray-100/5"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M9 5l7 7-7 7" class="hx-origin-center hx-transition-transform rtl:-hx-rotate-180"/></svg></span></a><div class="ltr:hx-pr-0 hx-overflow-hidden"><ul class='hx-relative hx-flex hx-flex-col hx-gap-1 before:hx-absolute before:hx-inset-y-1 before:hx-w-px before:hx-bg-gray-200 before:hx-content-[""] ltr:hx-ml-3 ltr:hx-pl-3 ltr:before:hx-left-0 rtl:hx-mr-3 rtl:hx-pr-3 rtl:before:hx-right-0 dark:before:hx-bg-neutral-800'><li class="hx-flex hx-flex-col"><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
hx-text-gray-500 hover:hx-bg-gray-100 hover:hx-text-gray-900 contrast-more:hx-border contrast-more:hx-border-transparent contrast-more:hx-text-gray-900 contrast-more:hover:hx-border-gray-900 dark:hx-text-neutral-400 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50 contrast-more:dark:hx-text-gray-50 contrast-more:dark:hover:hx-border-gray-50" href=/obitools4-doc/commands/alignments/obipairing/>Obipairing: align the forward and reverse paired reads</a></li><li class="hx-flex hx-flex-col open"><a class="hx-flex hx-items-center hx-justify-between hx-gap-2 hx-cursor-pointer hx-rounded hx-px-2 hx-py-1.5 hx-text-sm hx-transition-colors [-webkit-tap-highlight-color:transparent] [-webkit-touch-callout:none] [word-break:break-word]
sidebar-active-item hx-bg-primary-100 hx-font-semibold hx-text-primary-800 contrast-more:hx-border contrast-more:hx-border-primary-500 dark:hx-bg-primary-400/10 dark:hx-text-primary-600 contrast-more:dark:hx-border-primary-500" href=/obitools4-doc/commands/alignments/obipcr/>Obipcr: the electronic PCR tool</a></li></ul></div></li></ul></div><div class="hx-sticky hx-bottom-0 hx-bg-white dark:hx-bg-dark hx-mx-4 hx-py-4 hx-shadow-[0_-12px_16px_#fff] hx-flex hx-items-center hx-gap-2 dark:hx-border-neutral-800 dark:hx-shadow-[0_-12px_16px_#111] contrast-more:hx-border-neutral-400 contrast-more:hx-shadow-none contrast-more:dark:hx-shadow-none hx-border-t" data-toggle-animation=show><div class="hx-flex hx-grow hx-flex-col"><button title="Change theme" data-theme=light class="theme-toggle hx-group hx-h-7 hx-rounded-md hx-px-2 hx-text-left hx-text-xs hx-font-medium hx-text-gray-600 hx-transition-colors dark:hx-text-gray-400 hover:hx-bg-gray-100 hover:hx-text-gray-900 dark:hover:hx-bg-primary-100/5 dark:hover:hx-text-gray-50" type=button aria-label="Change theme"><div class="hx-flex hx-items-center hx-gap-2 hx-capitalize"><svg height="12" class="group-data-[theme=light]:hx-hidden" fill="none" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" aria-hidden="true"><path stroke-linecap="round" stroke-linejoin="round" d="M12 3v1m0 16v1m9-9h-1M4 12H3m15.364 6.364-.707-.707M6.343 6.343l-.707-.707m12.728.0-.707.707M6.343 17.657l-.707.707M16 12a4 4 0 11-8 0 4 4 0 018 0z"/></svg><span class="group-data-[theme=light]:hx-hidden">Light</span><svg height="12" class="group-data-[theme=dark]:hx-hidden" fill="none" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" aria-hidden="true"><path stroke-linecap="round" stroke-linejoin="round" d="M20.354 15.354A9 9 0 018.646 3.646 9.003 9.003.0 0012 21a9.003 9.003.0 008.354-5.646z"/></svg><span class="group-data-[theme=dark]:hx-hidden">Dark</span></div></button></div></div></aside><nav class="hextra-toc hx-order-last hx-hidden hx-w-64 hx-shrink-0 xl:hx-block print:hx-hidden hx-px-4" aria-label="table of contents"><div class="hextra-scrollbar hx-sticky hx-top-16 hx-overflow-y-auto hx-pr-4 hx-pt-6 hx-text-sm [hyphens:auto] hx-max-h-[calc(100vh-var(--navbar-height)-env(safe-area-inset-bottom))] ltr:hx--mr-4 rtl:hx--ml-4"><p class="hx-mb-4 hx-font-semibold hx-tracking-tight">On this page</p><ul><li class="hx-my-2 hx-scroll-my-6 hx-scroll-py-6"><a class="hx-font-semibold hx-inline-block hx-text-gray-500 hover:hx-text-gray-900 dark:hx-text-gray-400 dark:hover:hx-text-gray-300 contrast-more:hx-text-gray-900 contrast-more:hx-underline contrast-more:dark:hx-text-gray-50 hx-w-full hx-break-words" href=#description->Description :</a></li><li class="hx-my-2 hx-scroll-my-6 hx-scroll-py-6"><a class="ltr:hx-pl-4 rtl:hx-pr-4 hx-inline-block hx-text-gray-500 hover:hx-text-gray-900 dark:hx-text-gray-400 dark:hover:hx-text-gray-300 contrast-more:hx-text-gray-900 contrast-more:hx-underline contrast-more:dark:hx-text-gray-50 hx-w-full hx-break-words" href=#synopsis>Synopsis</a></li><li class="hx-my-2 hx-scroll-my-6 hx-scroll-py-6"><a class="ltr:hx-pl-4 rtl:hx-pr-4 hx-inline-block hx-text-gray-500 hover:hx-text-gray-900 dark:hx-text-gray-400 dark:hover:hx-text-gray-300 contrast-more:hx-text-gray-900 contrast-more:hx-underline contrast-more:dark:hx-text-gray-50 hx-w-full hx-break-words" href=#options>Options</a></li><li class="hx-my-2 hx-scroll-my-6 hx-scroll-py-6"><a class="ltr:hx-pl-8 rtl:hx-pr-8 hx-inline-block hx-text-gray-500 hover:hx-text-gray-900 dark:hx-text-gray-400 dark:hover:hx-text-gray-300 contrast-more:hx-text-gray-900 contrast-more:hx-underline contrast-more:dark:hx-text-gray-50 hx-w-full hx-break-words" href=#obipcr-mandatory-options>obipcr mandatory options:</a></li><li class="hx-my-2 hx-scroll-my-6 hx-scroll-py-6"><a class="ltr:hx-pl-8 rtl:hx-pr-8 hx-inline-block hx-text-gray-500 hover:hx-text-gray-900 dark:hx-text-gray-400 dark:hover:hx-text-gray-300 contrast-more:hx-text-gray-900 contrast-more:hx-underline contrast-more:dark:hx-text-gray-50 hx-w-full hx-break-words" href=#other-obipcr-specific-options>Other obipcr specific options:</a></li><li class="hx-my-2 hx-scroll-my-6 hx-scroll-py-6"><a class="ltr:hx-pl-4 rtl:hx-pr-4 hx-inline-block hx-text-gray-500 hover:hx-text-gray-900 dark:hx-text-gray-400 dark:hover:hx-text-gray-300 contrast-more:hx-text-gray-900 contrast-more:hx-underline contrast-more:dark:hx-text-gray-50 hx-w-full hx-break-words" href=#examples>Examples</a></li></ul><div class="hx-mt-8 hx-border-t hx-bg-white hx-pt-8 hx-shadow-[0_-12px_16px_white] dark:hx-bg-dark dark:hx-shadow-[0_-12px_16px_#111] hx-sticky hx-bottom-0 hx-flex hx-flex-col hx-items-start hx-gap-2 hx-pb-8 dark:hx-border-neutral-800 contrast-more:hx-border-t contrast-more:hx-border-neutral-400 contrast-more:hx-shadow-none contrast-more:dark:hx-border-neutral-400"><button aria-hidden=true id=backToTop onclick=scrollUp() class="hx-transition-all hx-duration-75 hx-opacity-0 hx-text-xs hx-font-medium hx-text-gray-500 hover:hx-text-gray-900 dark:hx-text-gray-400 dark:hover:hx-text-gray-100 contrast-more:hx-text-gray-800 contrast-more:dark:hx-text-gray-50">
<span>Scroll to top</span><svg fill="none" viewBox="0 0 24 24" stroke-width="1.5" stroke="currentcolor" class="hx-inline ltr:hx-ml-1 rtl:hx-mr-1 hx-h-3.5 hx-w-3.5 hx-border hx-rounded-full hx-border-gray-500 hover:hx-border-gray-900 dark:hx-border-gray-400 dark:hover:hx-border-gray-100 contrast-more:hx-border-gray-800 contrast-more:dark:hx-border-gray-50"><path stroke-linecap="round" stroke-linejoin="round" d="M4.5 15.75l7.5-7.5 7.5 7.5"/></svg></button></div></div></nav><article class="hx-w-full hx-break-words hx-flex hx-min-h-[calc(100vh-var(--navbar-height))] hx-min-w-0 hx-justify-center hx-pb-8 hx-pr-[calc(env(safe-area-inset-right)-1.5rem)]"><main class="hx-w-full hx-min-w-0 hx-max-w-6xl hx-px-6 hx-pt-4 md:hx-px-12"><div class="hx-mt-1.5 hx-flex hx-items-center hx-gap-1 hx-overflow-hidden hx-text-sm hx-text-gray-500 dark:hx-text-gray-400 contrast-more:hx-text-current"><div class="hx-whitespace-nowrap hx-transition-colors hx-min-w-[24px] hx-overflow-hidden hx-text-ellipsis hover:hx-text-gray-900 dark:hover:hx-text-gray-100"><a href=/obitools4-doc/commands/>Commands</a></div><svg class="hx-w-3.5 hx-shrink-0 rtl:-hx-rotate-180" fill="none" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" aria-hidden="true"><path stroke-linecap="round" stroke-linejoin="round" d="M9 5l7 7-7 7"/></svg><div class="hx-whitespace-nowrap hx-transition-colors hx-min-w-[24px] hx-overflow-hidden hx-text-ellipsis hover:hx-text-gray-900 dark:hover:hx-text-gray-100"><a href=/obitools4-doc/commands/alignments/>Sequence alignments</a></div><svg class="hx-w-3.5 hx-shrink-0 rtl:-hx-rotate-180" fill="none" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" aria-hidden="true"><path stroke-linecap="round" stroke-linejoin="round" d="M9 5l7 7-7 7"/></svg><div class="hx-whitespace-nowrap hx-transition-colors hx-font-medium hx-text-gray-700 contrast-more:hx-font-bold contrast-more:hx-text-current dark:hx-text-gray-100 contrast-more:dark:hx-text-current">Obipcr: the electronic PCR tool</div></div><div class=content><h1>Obipcr: the electronic PCR tool</h1><h2>Description :<span class="hx-absolute -hx-mt-20" id=description-></span>
<a href=#description- class=subheading-anchor aria-label="Permalink for this section"></a></h2><p><code>obipcr</code> is the successor of <a href=https://metabarcoding.org/ecopcr target=_blank rel=noopener><code>ecoPCR</code></a>. It is known as an electronic PCR software.</p><h3>Synopsis<span class="hx-absolute -hx-mt-20" id=synopsis></span>
<a href=#synopsis class=subheading-anchor aria-label="Permalink for this section"></a></h3><div class="hextra-code-block hx-relative hx-mt-6 first:hx-mt-0 hx-group/code"><div><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>obipcr --forward &lt;string&gt; --max-length|-L &lt;int&gt; --reverse &lt;string&gt;
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--allowed-mismatches|-e &lt;int&gt;<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--batch-size &lt;int&gt;<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--circular|-c<span style=color:#f92672>]</span>
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--compress|-Z<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--debug<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--delta|-D &lt;int&gt;<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--ecopcr<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--embl<span style=color:#f92672>]</span>
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--fasta<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--fasta-output<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--fastq<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--fastq-output<span style=color:#f92672>]</span>
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--force-one-cpu<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--fragmented<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--genbank<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--help|-h|-?<span style=color:#f92672>]</span>
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--input-OBI-header<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--input-json-header<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--json-output<span style=color:#f92672>]</span>
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--max-cpu &lt;int&gt;<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--min-length|-l &lt;int&gt;<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--no-order<span style=color:#f92672>]</span>
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--no-progressbar<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--only-complete-flanking<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--out|-o &lt;FILENAME&gt;<span style=color:#f92672>]</span>
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--output-OBI-header|-O<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--output-json-header<span style=color:#f92672>]</span>
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--paired-with &lt;FILENAME&gt;<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--pprof<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--pprof-goroutine &lt;int&gt;<span style=color:#f92672>]</span>
</span></span><span style=display:flex><span>       <span style=color:#f92672>[</span>--pprof-mutex &lt;int&gt;<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--skip-empty<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--solexa<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>--version<span style=color:#f92672>]</span> <span style=color:#f92672>[</span>&lt;args&gt;<span style=color:#f92672>]</span></span></span></code></pre></div></div><div class="hextra-code-copy-btn-container hx-opacity-0 hx-transition group-hover/code:hx-opacity-100 hx-flex hx-gap-1 hx-absolute hx-m-[11px] hx-right-0 hx-top-0"><button class="hextra-code-copy-btn hx-group/copybtn hx-transition-all active:hx-opacity-50 hx-bg-primary-700/5 hx-border hx-border-black/5 hx-text-gray-600 hover:hx-text-gray-900 hx-rounded-md hx-p-1.5 dark:hx-bg-primary-300/10 dark:hx-border-white/10 dark:hx-text-gray-400 dark:hover:hx-text-gray-50" title="Copy code"><div class="copy-icon group-[.copied]/copybtn:hx-hidden hx-pointer-events-none hx-h-4 hx-w-4"></div><div class="success-icon hx-hidden group-[.copied]/copybtn:hx-block hx-pointer-events-none hx-h-4 hx-w-4"></div></button></div></div><h3>Options<span class="hx-absolute -hx-mt-20" id=options></span>
<a href=#options class=subheading-anchor aria-label="Permalink for this section"></a></h3><h4><code>obipcr</code> mandatory options:<span class="hx-absolute -hx-mt-20" id=obipcr-mandatory-options></span>
<a href=#obipcr-mandatory-options class=subheading-anchor aria-label="Permalink for this section"></a></h4><ul><li><b><code class=language-bash>--forward</code></b>: The forward primer used for the electronic PCR. IUPAC codes can be used in the pattern.</li><li><b><code class=language-bash>--reverse</code></b>: The reverse primer used for the electronic PCR. IUPAC codes can be used in the pattern.</li><li><b><code class=language-bash>--max-length</code></b>
| <b><code class=language-bash>-L</code></b>
: Maximum length of the barcode, primers excluded.</li></ul><h4>Other <code>obipcr</code> specific options:<span class="hx-absolute -hx-mt-20" id=other-obipcr-specific-options></span>
<a href=#other-obipcr-specific-options class=subheading-anchor aria-label="Permalink for this section"></a></h4><ul><li><b><code class=language-bash>--allowed-mismatches</code></b>
| <b><code class=language-bash>-e</code></b>
: Maximum number of mismatches allowed for each primer (default: 0).</li><li><b><code class=language-bash>--min-length</code></b>
| <b><code class=language-bash>-l</code></b>
: Minimum length of the barcode primers excluded (default: no minimum length).</li><li><b><code class=language-bash>--circular</code></b>
| <b><code class=language-bash>-c</code></b>
: Considers that sequences are circular. (default: sequences are considered linear)</li><li><b><code class=language-bash>--delta</code></b>
| <b><code class=language-bash>-D</code></b>
: Without this option, only the barcode sequences will be output, without the priming sites. This option allows to add the priming sites and the flanking sequences of the priming sites over a length of <code>delta</code> to each side of the barcode.</li><li><b><code class=language-bash>--only-complete-flanking</code></b>: Works in conjunction with <strong>&ndash;delta</strong>. Prints only sequences with full-length flanking sequences (default: prints every sequence regardless of whether the flanking sequences are present).</li></ul><h4>Controlling the input data<span class="hx-absolute -hx-mt-20" id=controlling-the-input-data></span>
<a href=#controlling-the-input-data class=subheading-anchor aria-label="Permalink for this section"></a></h4><ul><li><b><code class=language-bash>--fasta</code></b>: Indicate that sequence data is following the FASTA format.</li><li><b><code class=language-bash>--fastq</code></b>: Indicate that sequence data is following the FASTQ format.</li><li><b><code class=language-bash>--embl</code></b>: Indicate that sequence data is following the EMBL flatfile format.</li><li><b><code class=language-bash>--genbank</code></b>: Indicate that sequence data is following the Genbank flatfile format.</li><li><b><code class=language-bash>--ecopcr</code></b>: Indicate that sequence data is following old ecoPCR tabulated format.</li><li><b><code class=language-bash>--input-OBI-header</code></b>: FASTA/FASTQ title line annotations follow the old OBI format.</li><li><b><code class=language-bash>--input-json-header</code></b>: FASTA/FASTQ title line annotations follow the json format.</li><li><b><code class=language-bash>--solexa</code></b>: Decodes quality string according to the Solexa specification. (default: the standard Sanger encoding is used, env: <strong>OBISSOLEXA</strong>)</li></ul><h4>Controlling the output data<span class="hx-absolute -hx-mt-20" id=controlling-the-output-data></span>
<a href=#controlling-the-output-data class=subheading-anchor aria-label="Permalink for this section"></a></h4><ul><li><b><code class=language-bash>--compress</code></b>
| <b><code class=language-bash>-Z</code></b>
: Output is compressed using gzip (default: false).</li><li><b><code class=language-bash>--no-order</code></b>: The <em>OBITools</em> ensure that the order between the input file and
the output file does not change. When multiple files are processed,
they are processed one at a time.
If the <strong>&ndash;no-order</strong> option is added to a command, multiple input
files can be opened at the same time and their contents processed
in parallel. This usually increases processing speed, but does not
guarantee the order of the sequences in the output file.
Also, processing multiple files in parallel may require more memory
to perform the computation.</li><li><b><code class=language-bash>--fasta-output</code></b>: Write sequence in fasta format (default if no quality data available).</li><li><b><code class=language-bash>--fastq-output</code></b>: Write sequence in fastq format (default if quality data available).</li><li><b><code class=language-bash>--json-output</code></b>: Write sequence in json format.</li><li><b><code class=language-bash>--out</code></b>
| <b><code class=language-bash>-o</code></b>
&lt;FILENAME>: Filename used for saving the output (default: &ldquo;-&rdquo;, the standard output)</li><li><b><code class=language-bash>--output-OBI-header</code></b>
| <b><code class=language-bash>-O</code></b>
: output FASTA/FASTQ title line annotations follow OBI format (default: JSON).</li><li><b><code class=language-bash>--output-json-header</code></b>: output FASTA/FASTQ title line annotations follow json format (the default format).</li><li><b><code class=language-bash>--skip-empty</code></b>: Sequences of length equal to zero are suppressed from the output (default: false).</li><li><b><code class=language-bash>--no-progressbar</code></b>: Disable the progress bar printing (default: false).</li></ul><h4>General options<span class="hx-absolute -hx-mt-20" id=general-options></span>
<a href=#general-options class=subheading-anchor aria-label="Permalink for this section"></a></h4><ul><li><b><code class=language-bash>--help</code></b>
| <b><code class=language-bash>-h|-?</code></b>
: Show this help.</li><li><b><code class=language-bash>--version</code></b>: Print the version and exits.</li></ul><h4>Computation related options<span class="hx-absolute -hx-mt-20" id=computation-related-options></span>
<a href=#computation-related-options class=subheading-anchor aria-label="Permalink for this section"></a></h4><ul><li><b><code class=language-bash>--max-cpu</code></b> &lt;INTEGER>: <em>OBITools</em> can take advantage of your computer&rsquo;s multi-core
architecture by parallelizing the computation across all available CPUs.
Computing on more CPUs usually requires more memory to perform the
computation. Reducing the number of CPUs used to perform a calculation
is also a way to indirectly control the amount of memory used by the
process. The number of CPUs used by <em>OBITools</em> can also be controlled
by setting the <strong>OBIMAXCPU</strong> environment variable.</li><li><b><code class=language-bash>--force-one-cpu</code></b>: Force to use only one CPU core for parallel processing (default: false).</li><li><b><code class=language-bash>--batch-size</code></b> &lt;INTEGER>: Number of sequence per batch for parallel processing (default: 1000, env: <strong>OBIBATCHSIZE</strong>)</li></ul><h4>Debug related options<span class="hx-absolute -hx-mt-20" id=debug-related-options></span>
<a href=#debug-related-options class=subheading-anchor aria-label="Permalink for this section"></a></h4><ul><li><b><code class=language-bash>--debug</code></b>: Enable debug mode, by setting log level to debug. (default: false, env: <strong>OBIDEBUG</strong>).</li><li><b><code class=language-bash>--pprof</code></b>: Enable pprof server. Look at the log for details. (default: false).</li><li><b><code class=language-bash>--pprof-mutex</code></b> &lt;INTEGER>: Enable profiling of mutex lock. (default: 10, env: <strong>OBIPPROFMUTEX</strong>)</li><li><b><code class=language-bash>--pprof-goroutine</code></b> &lt;INTEGER>: Enable profiling of goroutine blocking profile. (default: 6060, env: <strong>OBIPPROFGOROUTINE</strong>)</li></ul><h3>Examples<span class="hx-absolute -hx-mt-20" id=examples></span>
<a href=#examples class=subheading-anchor aria-label="Permalink for this section"></a></h3><p>The minimal <code>ecopcr</code> command looks like this:</p><div class="hextra-code-block hx-relative hx-mt-6 first:hx-mt-0 hx-group/code"><div><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>obipcr -L <span style=color:#ae81ff>220</span>  <span style=color:#ae81ff>\ </span> 
</span></span><span style=display:flex><span>       --forward GGGCAATCCTGAGCCAA <span style=color:#ae81ff>\ </span>  
</span></span><span style=display:flex><span>       --reverse CCATTGAGTCTCTGCACCTATC <span style=color:#ae81ff>\
</span></span></span><span style=display:flex><span><span style=color:#ae81ff></span>       /data/Genbank/Release_261 <span style=color:#ae81ff>\
</span></span></span><span style=display:flex><span><span style=color:#ae81ff></span>       &gt; Sper01_obipcr.fasta</span></span></code></pre></div></div><div class="hextra-code-copy-btn-container hx-opacity-0 hx-transition group-hover/code:hx-opacity-100 hx-flex hx-gap-1 hx-absolute hx-m-[11px] hx-right-0 hx-top-0"><button class="hextra-code-copy-btn hx-group/copybtn hx-transition-all active:hx-opacity-50 hx-bg-primary-700/5 hx-border hx-border-black/5 hx-text-gray-600 hover:hx-text-gray-900 hx-rounded-md hx-p-1.5 dark:hx-bg-primary-300/10 dark:hx-border-white/10 dark:hx-text-gray-400 dark:hover:hx-text-gray-50" title="Copy code"><div class="copy-icon group-[.copied]/copybtn:hx-hidden hx-pointer-events-none hx-h-4 hx-w-4"></div><div class="success-icon hx-hidden group-[.copied]/copybtn:hx-block hx-pointer-events-none hx-h-4 hx-w-4"></div></button></div></div><p>It fetches the sequences from the NCBI Genbank release 261 database stored in the <code>/data/Genbank/Release_261</code> directory. The output is stored in the <code>Sper01_obipcr.fasta</code> file. The primer pair is specified as <code>GGGCAATCCTGAGCCAA</code> and <code>CCATTGAGTCTCTGCACCTATC</code> using the <code>--forward</code> and <code>--reverse</code> options. These primers correspond to the marker <em>Sper01</em>. The <code>-L</code> option specifies the maximum length of the barcode, primers excluded, here 220 nucleotides. By default no mismatches are allowed between the primers and the priming sites.</p><p>To allows mismatches between the primers and the priming sites, use the <code>--allowed-mismatches</code> option or its short form <code>-e</code>. Here, the maximum number of mismatches allowed is 3. This maximum number of mismatches is allowed per primer. The error can occur at any position of the primer.</p><div class="hextra-code-block hx-relative hx-mt-6 first:hx-mt-0 hx-group/code"><div><div class=highlight><pre tabindex=0 style=color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4><code class=language-bash data-lang=bash><span style=display:flex><span>obipcr -e <span style=color:#ae81ff>3</span>  <span style=color:#ae81ff>\ </span> 
</span></span><span style=display:flex><span>       -L <span style=color:#ae81ff>220</span>  <span style=color:#ae81ff>\ </span> 
</span></span><span style=display:flex><span>       --forward GGGCAATCCTGAGCCAA <span style=color:#ae81ff>\
</span></span></span><span style=display:flex><span><span style=color:#ae81ff></span>       --reverse CCATTGAGTCTCTGCACCTATC <span style=color:#ae81ff>\
</span></span></span><span style=display:flex><span><span style=color:#ae81ff></span>       /data/Genbank/Release_261 <span style=color:#ae81ff>\
</span></span></span><span style=display:flex><span><span style=color:#ae81ff></span>       &gt; Sper01_obipcr.fasta</span></span></code></pre></div></div><div class="hextra-code-copy-btn-container hx-opacity-0 hx-transition group-hover/code:hx-opacity-100 hx-flex hx-gap-1 hx-absolute hx-m-[11px] hx-right-0 hx-top-0"><button class="hextra-code-copy-btn hx-group/copybtn hx-transition-all active:hx-opacity-50 hx-bg-primary-700/5 hx-border hx-border-black/5 hx-text-gray-600 hover:hx-text-gray-900 hx-rounded-md hx-p-1.5 dark:hx-bg-primary-300/10 dark:hx-border-white/10 dark:hx-text-gray-400 dark:hover:hx-text-gray-50" title="Copy code"><div class="copy-icon group-[.copied]/copybtn:hx-hidden hx-pointer-events-none hx-h-4 hx-w-4"></div><div class="success-icon hx-hidden group-[.copied]/copybtn:hx-block hx-pointer-events-none hx-h-4 hx-w-4"></div></button></div></div></div><div class=hx-mt-16></div><div class="hx-mb-8 hx-flex hx-items-center hx-border-t hx-pt-8 dark:hx-border-neutral-800 contrast-more:hx-border-neutral-400 dark:contrast-more:hx-border-neutral-400 print:hx-hidden"><a href=/obitools4-doc/commands/alignments/obipairing/ title="Obipairing: align the forward and reverse paired reads" class="hx-flex hx-max-w-[50%] hx-items-center hx-gap-1 hx-py-4 hx-text-base hx-font-medium hx-text-gray-600 hx-transition-colors [word-break:break-word] hover:hx-text-primary-600 dark:hx-text-gray-300 md:hx-text-lg ltr:hx-pr-4 rtl:hx-pl-4"><svg class="hx-inline hx-h-5 hx-shrink-0 ltr:hx-rotate-180" fill="none" viewBox="0 0 24 24" stroke-width="2" stroke="currentcolor" aria-hidden="true"><path stroke-linecap="round" stroke-linejoin="round" d="M9 5l7 7-7 7"/></svg>Obipairing: align the forward and reverse paired reads</a></div></main></article></div><footer class="hextra-footer hx-bg-gray-100 hx-pb-[env(safe-area-inset-bottom)] dark:hx-bg-neutral-900 print:hx-bg-transparent"><div class="hx-max-w-screen-xl hx-mx-auto hx-flex hx-justify-center hx-py-12 hx-pl-[max(env(safe-area-inset-left),1.5rem)] hx-pr-[max(env(safe-area-inset-right),1.5rem)] hx-text-gray-600 dark:hx-text-gray-400 md:hx-justify-start"><div class="hx-flex hx-w-full hx-flex-col hx-items-center sm:hx-items-start"><div class=hx-font-semibold><a class="hx-flex hx-text-sm hx-items-center hx-gap-1 hx-text-current" target=_blank rel="noopener noreferrer" title="Hextra GitHub Homepage" href=https://github.com/imfing/hextra><span>Powered by Hextra<svg height="1em" class="hx-inline-block ltr:hx-ml-1 rtl:hx-mr-1 hx-align-[-2.5px]" viewBox="0 0 180 180" fill="currentcolor"><path fill-rule="evenodd" clip-rule="evenodd" d="m105.50024 22.224647c-9.59169-5.537563-21.40871-5.537563-31.000093.0L39.054693 42.689119C29.463353 48.226675 23.55484 58.460531 23.55484 69.535642v40.928918c0 11.07542 5.908513 21.3092 15.499853 26.84652l35.445453 20.46446c9.591313 5.53732 21.408404 5.53732 31.000094.0l35.44507-20.46446c9.59131-5.53732 15.49985-15.7711 15.49985-26.84652V69.535642c0-11.075111-5.90854-21.308967-15.49985-26.846523zM34.112797 85.737639c-1.384445 2.397827-1.384445 5.352099.0 7.749927l24.781554 42.922974c1.38437 2.39783 3.942853 3.87496 6.711592 3.87496h49.563107c2.76905.0 5.3273-1.47713 6.71144-3.87496l24.78194-42.922974c1.38414-2.397828 1.38414-5.3521.0-7.749927L121.88049 42.814746c-1.38414-2.397828-3.94239-3.874964-6.71144-3.874964H65.605944c-2.768739.0-5.327223 1.477059-6.711592 3.874964z" style="stroke-width:.774993"/></svg></span></a></div></div></div></footer><script defer src=/obitools4-doc/js/main.min.998f7d6b2520d6d9875145cbe14b35964b365b44b3b1bfa6b797a74a9c173b07.js integrity="sha256-mY99ayUg1tmHUUXL4Us1lks2W0Szsb+mt5enSpwXOwc="></script><script defer src=/obitools4-doc/lib/flexsearch/flexsearch.bundle.min.0425860527cc9968f9f049421c7a56b39327d475e2e3a8f550416be3a9134327.js integrity="sha256-BCWGBSfMmWj58ElCHHpWs5Mn1HXi46j1UEFr46kTQyc="></script><script defer src=/obitools4-doc/en.search.min.6fba654e1f448aaf5ced9d57e2f0ae047079b52ee55913d0eded2698f9830908.js integrity="sha256-b7plTh9Eiq9c7Z1X4vCuBHB5tS7lWRPQ7e0mmPmDCQg="></script></body></html>